docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,283 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As more custom attestation plugins are developed and more features are requested, new APIs are added to:
+
+1. Query registered attesters and verifiers.
+2. Select a format of the evidence that the registered attesters can generate.
+
+The following API was affected:
+
+1. Built-in SGX attester and verifier plugin
+
+This document describes the added and affected APIs as well as keeping
+the existing support for version 2 of the API.  This document assumes the user is familiar with
+the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. In order to simplify evidence generation in attestation applications, change was made to split built-in SGX attester into 2 attesters - one for SGX report and the other for SGX quote. Similar plugin splitting applies to built-in SGX verifier.
+
+Local and remote attestation is a TEE-specific feature. Specifying LOCAL or REMOTE flag when generating evidence can be confusing to many attesters. These flags were deprecated.
+
+User Experience
+---------------
+
+The user experience is improved by the ability of querying what the registered attestation verifiers can verify and attesters can attest. This is a typical attestation scenario:
+
+- Verifier application
+    1. Registers appropriate verifier plugins (oe_register_verifier).
+    2. Gets the format IDs of registered verifier plugins (oe_get_registered_verifier_format_ids).
```

Comment:
It's a convenience function but can be helpful when plugins are registered/unregistered from different compilation unit or moment in the application.

---
In reply to: [387358601](https://github.com/openenclave/openenclave/pull/2621#discussion_r387358601) [](ancestors = 387358601)

Reply:
oe_get_registered_attester_format_ids() and oe_get_registered_verifier_format_ids() are removed since they are just convenience functions. They can be added in the future when needed.

---
In reply to: [389410611](https://github.com/openenclave/openenclave/pull/2621#discussion_r389410611) [](ancestors = 389410611,387358601)